File: 02_ImageFiltering/02_LinearFiltering/04_Laplacian Pyramid and Image Residuals/lpimageresiduals.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>

cv::Mat applyGaussianFilter(const cv::Mat& image, int kernelSize, double sigma) {
    cv::Mat result;
    cv::GaussianBlur(image, result, cv::Size(kernelSize, kernelSize), sigma);
    return result;
}

cv::Mat downsampleImage(const cv::Mat& image) {
    cv::Mat downsampled;
    cv::resize(image, downsampled, cv::Size(image.cols / 2, image.rows / 2));
    return downsampled;
}

cv::Mat upsampleImage(const cv::Mat& image, cv::Size size) {
    cv::Mat upsampled;
    cv::resize(image, upsampled, size);
    return upsampled;
}

std::vector<cv::Mat> createLaplacianPyramid(const cv::Mat& image, int levels) {
    std::vector<cv::Mat> gaussianPyramid;
    std::vector<cv::Mat> laplacianPyramid;
    
    // Construct Gaussian Pyramid
    gaussianPyramid.push_back(image);
    for (int i = 1; i < levels; ++i) {
        cv::Mat blurred = applyGaussianFilter(gaussianPyramid.back(), 5, 1.0);
        cv::Mat downsampled = downsampleImage(blurred);
        gaussianPyramid.push_back(downsampled);
    }
    
    // Construct Laplacian Pyramid
    for (int i = 0; i < levels - 1; ++i) {
        cv::Mat upsampled = upsampleImage(gaussianPyramid[i + 1], gaussianPyramid[i].size());
        cv::Mat laplacian = gaussianPyramid[i] - upsampled;
        laplacianPyramid.push_back(laplacian);
    }
    laplacianPyramid.push_back(gaussianPyramid.back()); // Smallest level remains unchanged
    
    return laplacianPyramid;
}

int main() {
    cv::Mat image = cv::imread("../../01_Images/Checker_original.png", cv::IMREAD_GRAYSCALE);
    if (image.empty()) {
        std::cerr << "Error: Image not found!" << std::endl;
        return -1;
    }
    
    std::vector<cv::Mat> laplacianPyramid = createLaplacianPyramid(image, 3);
    
    cv::imwrite("../../01_Images/laplacian_pyramid_level1.png", laplacianPyramid[0]);
    cv::imwrite("../../01_Images/laplacian_pyramid_level2.png", laplacianPyramid[1]);
    
    return 0;
}

File: 02_ImageFiltering/02_LinearFiltering/04_Laplacian Pyramid and Image Residuals/lpImageResidualsPy/lpimageresiduals.py
import cv2
import numpy as np

def apply_gaussian_filter(image, kernel_size, sigma):
    return cv2.GaussianBlur(image, (kernel_size, kernel_size), sigma)

def downsample_image(image):
    return cv2.resize(image, (image.shape[1] // 2, image.shape[0] // 2))

def upsample_image(image, size):
    return cv2.resize(image, size)

def create_laplacian_pyramid(image, levels):
    gaussian_pyramid = [image]
    laplacian_pyramid = []
    
    for _ in range(1, levels):
        blurred = apply_gaussian_filter(gaussian_pyramid[-1], 5, 1.0)
        downsampled = downsample_image(blurred)
        gaussian_pyramid.append(downsampled)
    
    for i in range(levels - 1):
        upsampled = upsample_image(gaussian_pyramid[i + 1], (gaussian_pyramid[i].shape[1], gaussian_pyramid[i].shape[0]))
        laplacian = cv2.subtract(gaussian_pyramid[i], upsampled)
        laplacian_pyramid.append(laplacian)
    
    laplacian_pyramid.append(gaussian_pyramid[-1])  # Smallest level remains unchanged
    return laplacian_pyramid

if __name__ == "__main__":
    image = cv2.imread("../../01_Images/Checker_original.png", cv2.IMREAD_GRAYSCALE)
    if image is None:
        print("Error: Image not found!")
        exit(1)
    
    laplacian_pyramid = create_laplacian_pyramid(image, 3)
    
    cv2.imwrite("../../01_Images/laplacian_pyramid_level1.png", laplacian_pyramid[0])
    cv2.imwrite("../../01_Images/laplacian_pyramid_level2.png", laplacian_pyramid[1])
    
    print("Laplacian pyramid generated successfully!")
